src/main/java/io/github/seggan/slimypowersuits/machines/ElementForge.java

```diff
@@ -0,0 +1,153 @@
+            List<ItemStack[]> inputs = RecipeType.getRecipeInputList(this);
+
+            for (ItemStack[] input : inputs) {
+                if (isCraftable(inv, input)) {
+                    ItemStack output = RecipeType.getRecipeOutputList(this, input).clone();
+
+                    if (Slimefun.hasUnlocked(p, output, true)) {
+                        Inventory outputInv = findOutputInventory(output, dispenser, inv);
+
+                        if (outputInv != null) {
+                            craft(output, inv, outputInv);
+
+                            movePiston(piston, piss, true);
+
+                            Utils.runSync(() -> movePiston(piston, piss, false), 10L);
+
+                        } else {
+                            SlimefunPlugin.getLocalization().sendMessage(p, "machines.full-inventory", true);
+                        }
+                    }
+
+                    return;
+                }
+            }
+
+            SlimefunPlugin.getLocalization().sendMessage(p, "machines.pattern-not-found", true);
+        }
+    }
+
+    private boolean isCraftable(Inventory inv, ItemStack[] recipe) {
+        for (int j = 0; j < inv.getContents().length; j++) {
+            if (!SlimefunUtils.isItemSimilar(inv.getContents()[j], recipe[j], true)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private void craft(ItemStack output, Inventory inv, Inventory outputInv) {
+        for (int j = 0; j < 9; j++) {
+            ItemStack item = inv.getContents()[j];
+
+            if (item != null && item.getType() != Material.AIR) {
+                ItemUtils.consumeItem(item, true);
+            }
+        }
+        outputInv.addItem(output);
+    }
+
+    private Block locateDispenser(Block b) {
+        Block block = null;
+
+        if (b.getRelative(1, 0, 0).getType() == Material.DISPENSER) {
+            block = b.getRelative(1, 0, 0);
+        } else if (b.getRelative(0, 0, 1).getType() == Material.DISPENSER) {
+            block = b.getRelative(0, 0, 1);
+        } else if (b.getRelative(-1, 0, 0).getType() == Material.DISPENSER) {
+            block = b.getRelative(-1, 0, 0);
+        } else if (b.getRelative(0, 0, -1).getType() == Material.DISPENSER) {
+            block = b.getRelative(0, 0, -1);
+        }
+
+        return block;
+    }
+
+    private void movePiston(Block piston, Piston piss, boolean extended) {
+        piss.setExtended(extended);
+        piston.setBlockData(piss, false);
+
+        // Updating the Piston Head
+        if (extended) {
+            PistonHead head = (PistonHead) Material.PISTON_HEAD.createBlockData();
+            head.setFacing(BlockFace.DOWN);
+
+            piston.getRelative(BlockFace.DOWN).setBlockData(head, false);
+            piston.getWorld().playSound(piston.getLocation(), Sound.BLOCK_PISTON_CONTRACT, 0.5f, 0.5f);
+        } else {
+            piston.getRelative(BlockFace.DOWN).setType(Material.AIR);
```

Comment:
not sure how much better it would be but BlockFace.DOWN in a var?

Reply:
I think this wouldnt matter as enum values are stored in one location

Reply:
How much does this help performance wise? I feel like doing such a change would make it messier, not neater

Reply:
> 
> 
> I think this wouldnt matter as enum values are stored in one location